pkg/controllers/machine/lifecycle/launch.go

```diff
@@ -0,0 +1,147 @@
+
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+)
+
+type Launch struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cache         *cache.Cache // exists due to eventual consistency on the cache
+}
+
+func (l *Launch) Reconcile(ctx context.Context, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+
+	var err error
+	var created *v1alpha5.Machine
+
+	if ret, ok := l.cache.Get(client.ObjectKeyFromObject(machine).String()); ok {
+		created = ret.(*v1alpha5.Machine)
+	} else if _, ok := machine.Annotations[v1alpha5.MachineLinkedAnnotationKey]; ok {
+		created, err = l.linkMachine(ctx, machine)
+	} else {
+		created, err = l.launchMachine(ctx, machine)
+	}
+	// Either the machine launch/linking failed or the machine was deleted due to InsufficientCapacity/NotFound
+	if err != nil || created == nil {
+		return reconcile.Result{}, err
+	}
+	l.cache.SetDefault(client.ObjectKeyFromObject(machine).String(), created)
+	PopulateMachineDetails(machine, created)
+	machine.StatusConditions().MarkTrue(v1alpha5.MachineLaunched)
+	return reconcile.Result{}, nil
+}
+
+func (l *Launch) linkMachine(ctx context.Context, machine *v1alpha5.Machine) (*v1alpha5.Machine, error) {
+	created, err := l.cloudProvider.Get(ctx, machine.Annotations[v1alpha5.MachineLinkedAnnotationKey])
+	if err != nil {
+		if cloudprovider.IsMachineNotFoundError(err) {
+			if err = l.kubeClient.Delete(ctx, machine); err != nil {
+				return nil, client.IgnoreNotFound(err)
+			}
+			logging.FromContext(ctx).Debugf("garbage collected machine with no cloudprovider representation")
+			metrics.MachinesTerminatedCounter.With(prometheus.Labels{
+				metrics.ReasonLabel:      "garbage_collected",
+				metrics.ProvisionerLabel: machine.Labels[v1alpha5.ProvisionerNameLabelKey],
+			}).Inc()
+			return nil, nil
+		}
+		machine.StatusConditions().MarkFalse(v1alpha5.MachineLaunched, "LinkFailed", truncateMessage(err.Error()))
+		return nil, fmt.Errorf("linking machine, %w", err)
+	}
+	logging.FromContext(ctx).Debugf("linked machine")
+	return created, nil
+}
+
+func (l *Launch) launchMachine(ctx context.Context, machine *v1alpha5.Machine) (*v1alpha5.Machine, error) {
+	instanceTypeRequirement, _ := lo.Find(machine.Spec.Requirements, func(req v1.NodeSelectorRequirement) bool { return req.Key == v1.LabelInstanceTypeStable })
+	logging.FromContext(ctx).With("pods", machine.Spec.Resources.Requests[v1.ResourcePods],
+		"resources", machine.Spec.Resources.Requests, "instance-types", instanceTypeList(instanceTypeRequirement.Values)).Infof("launching machine")
+	created, err := l.cloudProvider.Create(ctx, machine)
+	if err != nil {
+		if cloudprovider.IsInsufficientCapacityError(err) {
```

Comment:
nit on similar comment from above on short circuiting some of the code so it's less nested.